### ISSUE TITLE
container restart after kill dockerd although configured with --live-restore flag

### ISSUE THREAD

**Issue by ryanlyy** [moby/moby#27057]
I am testing --live-restore flag. And using pkill -9 dockerd to simulate docker corrupt. I found: after kill, container is still live which is expected according to live-restore flag user manual. But when I do systemctl start dockerd, I found container is restarted and pid is changed too which seems is not expected. could you please help if there is configuration I missed?

Thanks and Best Regards,
Ryan

Environment:
1. Docker Version: 1.12.1
2. OS:  Centos 7 with 3.10.0-327.el7.x86_64

**Steps to reproduce the issue:**
1. Check pid of container: ps afx
2. pkill -9 dockerd
3. systemctl start docker
4. Check new PID of container: ps afx

**Describe the results you received:**
1. Before start docker, container works well (i.e: process is still running)
2. After start, container seems restart and process id is changed too

**Describe the results you expected:**
container not restart and pid not change, only this there is no impact on service when dockerd corrupt.

**Additional information you deem important (e.g. issue happens only occasionally):**
It can be reproduced
1. before pkill:
 6875 ?        Ssl    0:03 /usr/bin/dockerd -D
 6878 ?        Ssl    0:00  _ docker-containerd -l unix:///var/run/docker/libcontainerd/docker-con
 7055 ?        Sl     0:00  |   _ docker-containerd-shim b5cea0be1b65726da31fe9a52139f91384c613d51
 7066 pts/1    Ss+    0:00  |   |   _ bash
 7104 ?        Sl     0:00  |   _ docker-containerd-shim 308213aac0261f6edcb4c3038ef7af43eaa233289
 7114 pts/3    Ss+    0:00  |       _ nginx: master process nginx -g daemon off;
 7138 pts/3    S+     0:00  |           _ nginx: worker process
 7093 ?        Sl     0:00  _ /usr/bin/docker-proxy -proto tcp -host-ip 0.0.0.0 -host-port 32768 -
 7100 ?        Sl     0:00  _ /usr/bin/docker-proxy -proto tcp -host-ip 0.0.0.0 -host-port 32769 -
[root@cluster-etcd3 ~]#
1. After pkill and start docker:
   7231 ?        Ssl    0:00 /usr/bin/dockerd -D
   7234 ?        Ssl    0:00  _ docker-containerd -l unix:///var/run/docker/libcontainerd/docker-containerd.sock --s
   7395 ?        Sl     0:00  |   _ docker-containerd-shim b5cea0be1b65726da31fe9a52139f91384c613d5183cb4065950f6bbf
   7406 pts/1    Ss+    0:00  |   |   _ bash
   7444 ?        Sl     0:00  |   _ docker-containerd-shim 308213aac0261f6edcb4c3038ef7af43eaa2332897850951c4b4fc429
   7456 pts/3    Ss+    0:00  |       _ nginx: master process nginx -g daemon off;
   7478 pts/3    S+     0:00  |           _ nginx: worker process
   7433 ?        Sl     0:00  _ /usr/bin/docker-proxy -proto tcp -host-ip 0.0.0.0 -host-port 32768 -container-ip 172
   7440 ?        Sl     0:00  _ /usr/bin/docker-proxy -proto tcp -host-ip 0.0.0.0 -host-port 32769 -container-ip 172
   [root@cluster-etcd3 ~]#

**Output of `docker version`:**
[root@cluster-etcd3 ~]# docker version
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:
 OS/Arch:      linux/amd64

**Output of `docker info`:**
[root@cluster-etcd3 ~]# docker info
Containers: 2
 Running: 2
 Paused: 0
 Stopped: 0
Images: 5
Server Version: 1.12.1
Storage Driver: devicemapper
 Pool Name: dockervg-thinpool
 Pool Blocksize: 524.3 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file:
 Metadata file:
 Data Space Used: 1.064 GB
 Data Space Total: 8.154 GB
 Data Space Available: 7.09 GB
 Metadata Space Used: 352.3 kB
 Metadata Space Total: 83.89 MB
 Metadata Space Available: 83.53 MB
 Thin Pool Minimum Free Space: 815.3 MB
 Udev Sync Supported: true
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Library Version: 1.02.107-RHEL7 (2015-10-14)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null overlay bridge host
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 3.10.0-327.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 993 MiB
Name: cluster-etcd3
ID: 5AKB:77HJ:MFS4:NE36:G3WV:PC3I:CZXN:PEEU:7IWY:X543:72T3:HDKE
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 22
 Goroutines: 31
 System Time: 2016-09-30T01:34:47.769356423-04:00
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Insecure Registries:
 127.0.0.0/8
[root@cluster-etcd3 ~]#

**Additional environment details (AWS, VirtualBox, physical, etc.):**
test is based on VirtualBox

**Comment by thaJeztah** [comment]
Can you check your systemd unit file? Wondering if it's related to https://github.com/docker/docker/issues/21933, which was fixed by https://github.com/docker/docker/pull/23636 (https://github.com/docker/docker/pull/23636/files), and if you have those changes in your unit-file.

**Comment by ryanlyy** [comment]
Thanks thaJeztah, Sorry for late blocked by other issues. I just checked unit file: KillMode=process is configured.

TimeoutStartSec=0
# set delegate yes so that systemd does not reset the cgroups of docker containers
Delegate=yes
# kill only the docker process, not all processes in the cgroup
KillMode=process

[Install]
WantedBy=multi-user.target

Any comments?

**Comment by thaJeztah** [comment]
no, not directly 


@tonistiigi @crosbymichael any ideas?

**Comment by tonistiigi** [comment]
fwiw I tested live-restore with these systemd options and didn't have issues. We need daemon logs have a better idea about what may be going on here. Also, may be related to a specific image and how it handles stdio closing or stdout/stderr blocking.

**Comment by tonistiigi** [comment]
@ryanlyy In your logs I see ` /usr/bin/dockerd -D` so no `--live-restore`. This flag needs to be set on dockerd process both before and after restart. If this flag is not set then the behavior you are describing is correct.

**Comment by ryanlyy** [comment]
I configured them in daemon.json
[root@cluster-etcd3 docker]# cat daemon.json
{
        "storage-driver": "devicemapper",
        "storage-opts": [
                "dm.thinpooldev=/dev/mapper/dockervg-thinpool",
                "dm.basesize=50G",
                "dm.fs=xfs",
                "dm.use_deferred_removal=true",
                "dm.use_deferred_deletion=true",
                "dm.min_free_space=10%"
        ],
        "live-restore": true,

**Comment by ryanlyy** [comment]
Any comment? Does Live-restore work well? 

Thanks,
Ryan

**Comment by thaJeztah** [comment]
ping @tonistiigi PTAL

**Comment by crosbymichael** [comment]
@ryanlyy what version of systemd do you have install?  

You can check via `systemd --version`

**Comment by ryanlyy** [comment]
[root@cluster-misc ~]# /usr/lib/systemd/systemd --version
systemd 219
+PAM +AUDIT +SELINUX +IMA -APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ -LZ4 -SECCOMP +BLKID +ELFUTILS +KMOD +IDN
[root@cluster-misc ~]#

**Comment by thaJeztah** [comment]
ping @crosbymichael PTAL ^^

**Comment by marcbachmann** [comment]
I'm having the same issue with systemd, but I guess in my case it's an issue with the setup.
That is the error I got:
```
Aug 19 16:42:20 server docker[22293]: time="2017-08-19T16:42:20Z" level=error msg="error getting events from daemon: unexpected EOF"
```

So it looks like the docker client is exiting because the docker daemon stopped. systemd then restarts the service. Because of the restart this did the whole bootstrapping again and removed the container.

```
[Unit]
Description=HTTP Proxy
After=rancher-server.service

[Service]
Restart=always
TimeoutStartSec=0
RestartSec=30
ExecStartPre=-/bin/mkdir -p /tmp/caddy && /bin/chmod 777 /tmp/caddy
ExecStartPre=-/usr/bin/docker stop rancher-caddy
ExecStartPre=-/usr/bin/docker rm rancher-caddy
ExecStartPre=-/usr/bin/docker pull abiosoft/caddy:0.9.2
ExecStart=/usr/bin/docker run --network rancher --name rancher-caddy -v /etc/caddy:/etc/caddy -v /tmp/caddy:/home/caddy/.caddy -p 80:80 -p 443:443 abiosoft/caddy:0.9.2 --conf /etc/caddy/Caddyfile
``` 

I guess I'm better off without using systemd for services 😄

**Comment by thaJeztah** [comment]
Yes, trying to stay attached to a container won't work, because killing the daemon kills the API, so the client doesn't have anything to connect to

**Comment by tnguyenns1** [comment]
@thaJeztah So I'm seeing the same issue with docker-ce 18.03.0, in which restarting dockerd daemon spawned new PID. I'm using upstart.

Reproducible is almost the same as the origion post.

Here's daemon.json config:
`# cat /etc/docker/daemon.json
{
        "live-restore": true,
        "log-opts": {
                "mode": "non-blocking",
                "max-buffer-size": "4m"
        }
}`

**Comment by ryanlyy** [comment]
any fix submitted for this issue in later release?

Thanks,
Ryan

**Comment by sam-thibault** [comment]
I don't see any recent activity on this issue.  I will close it as stale.
